src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+
+        else:
+            mbox_id = self.mbox_uuid
+            wrapper.set_mbox_uuid(mbox_id)
+            wrapper.set_flags(flags)
+            wrapper.set_tags(tags)
+            wrapper.set_date(date)
+
+        def insert_mdoc_id(_, wrapper):
+            doc_id = wrapper.mdoc.doc_id
+            return self.mbox_indexer.insert_doc(
+                self.mbox_uuid, doc_id)
+
+        d = wrapper.create(self.store)
+        d.addCallback(insert_mdoc_id, wrapper)
+        d.addErrback(lambda f: f.printTraceback())
+
+        return d
+
+    def copy_msg(self, msg, newmailbox):
+        """
+        Copy the message to another collection. (it only makes sense for
+        mailbox collections)
+        """
+        if not self.is_mailbox_collection():
+            raise NotImplementedError()
+
+        def insert_copied_mdoc_id(wrapper):
+            return self.mbox_indexer.insert_doc(
+                newmailbox_uuid, wrapper.mdoc.doc_id)
+
+        wrapper = msg.get_wrapper()
+        d = wrapper.copy(self.store, newmailbox)
+        d.addCallback(insert_copied_mdoc_id)
+        return d
+
+    def delete_msg(self, msg):
+        """
+        Delete this message.
+        """
+        wrapper = msg.get_wrapper()
+
+        def delete_mdoc_id(_, wrapper):
+            doc_id = wrapper.mdoc.doc_id
+            return self.mbox_indexer.delete_doc_by_hash(
+                self.mbox_name, doc_id)
+        d = wrapper.delete(self.store)
+        d.addCallback(delete_mdoc_id, wrapper)
+        return d
+
+    def delete_all_flagged(self):
+        """
+        Delete all messages flagged as \\Deleted.
+        Used from IMAPMailbox.expunge()
+        """
+        def get_uid_list(hashes):
+            d = []
+            for h in hashes:
+                d.append(self.mbox_indexer.get_uid_from_doc_id(
+                         self.mbox_uuid, h))
+            return defer.gatherResults(d), hashes
+
+        def delete_uid_entries((uids, hashes)):
+            d = []
+            for h in hashes:
+                d.append(self.mbox_indexer.delete_doc_by_hash(
+                         self.mbox_uuid, h))
+            return defer.gatherResults(d).addCallback(
+                lambda _: uids)
+
+        mdocs_deleted = self.adaptor.del_all_flagged_messages(
+            self.store, self.mbox_uuid)
+        mdocs_deleted.addCallback(get_uid_list)
+        mdocs_deleted.addCallback(delete_uid_entries)
+        return mdocs_deleted
+
+    # TODO should add a delete-by-uid to collection?
+
+    def delete_all_docs(self):
+        def del_all_uid(uid_list):
+            deferreds = []
+            for uid in uid_list:
+                d = self.get_message_by_uid(uid)
+                d.addCallback(lambda msg: msg.delete())
+                deferreds.append(d)
+            return defer.gatherResults(deferreds)
+
+        d = self.all_uid_iter()
+        d.addCallback(del_all_uid)
+        return d
+
+    def _update_flags_or_tags(self, old, new, mode):
```

Comment:
Again my OCD on clean code. Can we move this method below the ones that used it?